### PR TITLE
[TACHYON-1520] Refactor TachyonWorker and BlockWorker

### PR DESCRIPTION
**Pull request by apc999**
https://tachyon.atlassian.net/browse/TACHYON-1520
So TachyonWorker becomes the real identity to start/stop different workers and servicies

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 1]
```diff
@@ -15,20 +15,108 @@
...
 
+  private TachyonConf mTachyonConf;
+  // The workers
+  /** The worker serving blocks */
+  private BlockWorker mBlockWorker;
+  /** The worker serving file system operations */
+  private FileSystemWorker mFileSystemWorker;
+
+  /** is true if the worker is serving the RPC server */
+  private boolean mIsServing = false;
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Rename this to `mIsServingRPC` to be more specific

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 2]
```diff
@@ -15,20 +15,108 @@
...
+  /** RPC local port for thrift */
+  private int mRPCPort;
+  /** The address for the rpc server */
+  private InetSocketAddress mWorkerAddress;
+  /** Net address of this worker */
+  private WorkerNetAddress mWorkerNetAddress;
+  /** Worker start time in milliseconds */
+  private long mStartTimeMs;
+
+  public TachyonWorker() {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: javadoc

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 3]
```diff
@@ -15,20 +15,108 @@
...
+      // Reset worker RPC port based on assigned port number
+      mTachyonConf.set(Constants.WORKER_RPC_PORT, Integer.toString(mRPCPort));
+      mThriftServer = createThriftServer();
+
+      mWorkerAddress =
+          NetworkAddressUtils.getConnectAddress(NetworkAddressUtils.ServiceType.WORKER_RPC,
+              mTachyonConf);
+
+    } catch (Exception e) {
+      LOG.error(e.getMessage(), e);
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Log some specific message?

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 4]
```diff
@@ -37,37 +125,195 @@
...
-      worker = new BlockWorker();
-       // Setup the file worker
-      LOG.info("Started file system worker at worker with id {}", WorkerIdRegistry.getWorkerId());
-      fileWorker = new FileSystemWorker(worker.getBlockDataManager());
-
+      worker = new TachyonWorker();
+      worker.start();
     } catch (Exception e) {
-      LOG.error("Failed to initialize the block worker, exiting.", e);
+      LOG.error("Uncaught exception while running worker, shutting down and exiting.", e);
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Do you need to call `stop()`?

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 5]
```diff
@@ -37,37 +125,195 @@
...
+   * @return the worker RPC service port
+   */
+  public int getRPCLocalPort() {
+    return mRPCPort;
+  }
+
+  /**
+   * @return the worker data service bind host (used by unit test only)
+   */
+  public String getDataBindHost() {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Rename this to `getBlockWorkerDataBindHost`?

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 6]
```diff
@@ -37,37 +125,195 @@
...
+    return mBlockWorker.getDataBindHost();
+  }
+
+  /**
+   * @return the worker data service port (used by unit test only)
+   */
+  public int getDataLocalPort() {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: ditto

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 7]
```diff
@@ -37,37 +125,195 @@
...
+    return mBlockWorker.getDataLocalPort();
+  }
+
+  /**
+   * @return the worker web service bind host (used by unit test only)
+   */
+  public String getWebBindHost() {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: ditto

File: servers/src/main/java/tachyon/worker/Worker.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package tachyon.worker;
+
+import java.io.IOException;
+
+/**
+ * Interface of a Tachyon worker.
+ */
+public interface Worker {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/Worker.java]: Maybe rename this `Server` so that both master and worker can implement?

**Comment by yupeng9** [comment]
Thanks for this refactoring. I left some comments.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 5 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: it is not decided yet whether dataserver will just belong to blockworker or not. very likely we will move dataserver from block worker to tachyon worker.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 6 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: explained above

File: servers/src/main/java/tachyon/worker/Worker.java (hunk 8 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/Worker.java]: there should be another interface `Server` added, and `Worker` and `Master` interface both extend. but that belongs to another PR maybe.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 1 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: done

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 2 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: added

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 3 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: added

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 4 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: added

**Comment by apc999** [comment]
@yupeng9  fixed your comments or explained. 

PTAL, Thanks

**Comment by yupeng9** [comment]
LGTM

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 9]
```diff
@@ -15,20 +15,111 @@
...
+      mTachyonConf.set(Constants.WORKER_RPC_PORT, Integer.toString(mRPCPort));
+      mThriftServer = createThriftServer();
+
+      mWorkerAddress =
+          NetworkAddressUtils.getConnectAddress(NetworkAddressUtils.ServiceType.WORKER_RPC,
+              mTachyonConf);
+
+    } catch (Exception e) {
+      LOG.error("Failed to initialize {}", this.getClass().getName(), e);
+      System.exit(-1);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Do we need the exception handling here, since it will be handled in the main method?

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 10]
```diff
@@ -37,37 +128,202 @@
...
-
+      worker = new TachyonWorker();
+      worker.start();
     } catch (Exception e) {
-      LOG.error("Failed to initialize the block worker, exiting.", e);
+      LOG.error("Uncaught exception while running worker, shutting down and exiting.", e);
+      if (worker != null) {
+        try {
+          worker.stop();
+        } catch (Exception ex) {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: `Exception e`

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 9 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: I was trying to keep it consistent with the style of `TachyonMaster`. Once there is a PR to make separate main classes, we could address that using a consistent style there for both worker and master?

Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: for now I simply remove the try-catch. let's wait another PR to make styles consistent.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 10 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: added.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 9 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: for now I move `new TachyonWorker()` out of try-catch in main.

**Comment by apc999** [comment]
@calvinjia PTAL

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 11]
```diff
@@ -37,37 +128,198 @@
...
+    // Start each worker
+    // Requirement: NetAddress set in WorkerContext, so block worker can initialize BlockMasterSync
+    // Consequence: worker id is granted
+    startWorkers();
+    LOG.info("Started worker with id {}", WorkerIdRegistry.getWorkerId());
+
+    mIsServingRPC = true;
+
+    // Start serving RPC, this will block
+    LOG.info("Tachyon Worker version {} started @ {} {}", Version.VERSION, mWorkerAddress);
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: there is an extra "{}"?

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 12]
```diff
@@ -37,37 +128,198 @@
...
+   * Stops the Tachyon worker server. Should only be called by tests.
+   */
+  public void stop() throws Exception {
+    if (mIsServingRPC) {
+      LOG.info("Stopping RPC server on Tachyon Worker @ {}", mWorkerAddress);
+      stopServing();
+      stopWorkers();
+      mIsServingRPC = false;
+    } else {
+      LOG.info("Stopping Tachyon Master @ {}", mWorkerAddress);
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Master -> Worker

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java [hunk 13]
```diff
@@ -286,62 +196,12 @@ public void stop() throws IOException {
...
-    try {
-      mWebServer.shutdownWebServer();
-    } catch (Exception e) {
-      LOG.error("Failed to stop web server", e);
-    }
+
     mBlockDataManager.stop();
-    while (!mDataServer.isClosed() || mThriftServer.isServing()) {
-      // The reason to stop and close again is due to some issues in Thrift.
+    while (!mDataServer.isClosed()) {
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: Do we understand why we need to close() more then once? If so, there should be a comment explaining it. If not, we should add a TODO to investigate further (and maybe a guess?).

File: servers/src/main/java/tachyon/worker/netty/NettyDataServer.java [hunk 14]
```diff
@@ -71,7 +71,7 @@ public NettyDataServer(final InetSocketAddress address, final BlockDataManager d
   public void close() throws IOException {
     int quietPeriodSecs = mTachyonConf.getInt(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_QUIET_PERIOD);
     int timeoutSecs = mTachyonConf.getInt(Constants.WORKER_NETWORK_NETTY_SHUTDOWN_TIMEOUT);
-    mChannelFuture.channel().close().awaitUninterruptibly();
+    mChannelFuture.channel().close().awaitUninterruptibly(timeoutSecs, TimeUnit.SECONDS);
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/netty/NettyDataServer.java]: if this is 0 seconds (in integration tests), will the the channel even be closed?

**Comment by gpang** [comment]
I left some comments. Thanks for this organization!

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 11 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: removed.

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 12 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: done

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java (hunk 13 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: TODO added.

File: servers/src/main/java/tachyon/worker/netty/NettyDataServer.java (hunk 14 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/netty/NettyDataServer.java]: left a TODO to investigate this issue.

Integration tests can complete successfully on my machine after this change.

**Comment by apc999** [comment]
@gpang I addressed your comments (mostly), PTAL

**Comment by gpang** [comment]
LGTM

**Comment by apc999** [comment]
Jenkins, test this please

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java [hunk 15]
```diff
@@ -286,62 +196,14 @@ public void stop() throws IOException {
...
-    try {
-      mWebServer.shutdownWebServer();
-    } catch (Exception e) {
-      LOG.error("Failed to stop web server", e);
-    }
+
     mBlockDataManager.stop();
-    while (!mDataServer.isClosed() || mThriftServer.isServing()) {
-      // The reason to stop and close again is due to some issues in Thrift.
+    // TODO(binfan): investigate why we need to close dataserver again. There usd to be a comment
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: s/usd/used

File: servers/src/main/java/tachyon/worker/WorkerContext.java [hunk 16]
```diff
@@ -30,6 +31,14 @@ private WorkerContext() {} // to prevent initialization
   private static TachyonConf sTachyonConf = new TachyonConf();
 
   /**
+   * The {@link WorkerSource} for collecting master metrics.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: worker metrics

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 17]
```diff
@@ -15,20 +15,111 @@
...
+  /**
+   * Constructor of {@link TachyonWorker}.
+   */
+  public TachyonWorker() {
+    try {
+      mStartTimeMs = System.currentTimeMillis();
+      mTachyonConf = WorkerContext.getConf();
+
+      mBlockWorker = new BlockWorker();
+      // Setup the file worker
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: either remove this comment or move it one line up and change it to `Setup individual workers`

**Comment by calvinjia** [comment]
LGTM

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java (hunk 15 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: done

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 18]
```diff
@@ -15,20 +15,111 @@
...
 
+import com.google.common.base.Throwables;
+
 import tachyon.Constants;
+import tachyon.Version;
+import tachyon.conf.TachyonConf;
+import tachyon.metrics.MetricsSystem;
+import tachyon.security.authentication.AuthenticationUtils;
+import tachyon.thrift.BlockWorkerClientService;
+import tachyon.thrift.WorkerNetAddress;
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: should we be using tachyon.worker.NetAddress instead?

File: servers/src/main/java/tachyon/worker/WorkerContext.java (hunk 16 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/WorkerContext.java]: done

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 17 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: removed

File: servers/src/main/java/tachyon/worker/TachyonWorker.java [hunk 19]
```diff
@@ -37,37 +128,198 @@
...
+    mIsServingRPC = true;
+
+    // Start serving RPC, this will block
+    LOG.info("Tachyon Worker version {} started @ {}", Version.VERSION, mWorkerAddress);
+    mThriftServer.serve();
+    LOG.info("Tachyon Worker version {} ended @ {}", Version.VERSION, mWorkerAddress);
+  }
+
+  /**
+   * Stops the Tachyon worker server. Should only be called by tests.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: can we instead use Whitebox to restrict the access?

**Comment by jsimsa** [comment]
@apc999 left some comments

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 18 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: since this is a refactor PR, we should probably limit the scope of this PR by just making it moving code from one class to the other. 

Using `WorkerNetAddress` requires functionality change of code. We could probably address this in separate PRs (if necessary).

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 19 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: The comment on "Should only be called by tests." is obsolete now as we also call this method in main. Removed it from javadoc

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 18 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: SGTM

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 19 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Could you remove the same comment from TachyonMaster.java? thanks

**Comment by jsimsa** [comment]
@apc999 just one more minor comment, thanks for the great refactor ... LGTM

File: servers/src/main/java/tachyon/worker/TachyonWorker.java (hunk 19 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/TachyonWorker.java]: Update the javadoc in TachyonMaster, although it is not called by TachyonMaster main, but by fault-tolerant mater.

**Comment by apc999** [comment]
@jsimsa thanks for the comment. I have addressed it.

**Comment by jsimsa** [comment]
LGTM
